lib/astutils.cpp

```diff
@@ -133,9 +133,74 @@ const Token * astIsVariableComparison(const Token *tok, const std::string &comp,
     return ret;
 }
 
+const Token * getVariableInitExpression(const Variable * var)
+{
+    if(!var || !var->declEndToken())
+        return nullptr;
+    if(Token::Match(var->declEndToken(), "; %varid% =", var->declarationId()))
+        return var->declEndToken()->tokAt(2)->astOperand2();
+    return var->declEndToken()->astOperand2();
+}
+
+bool isInLoopCondition(const Token * tok) {
+    return Token::Match(tok->astTop()->previous(), "for|while (");
+}
+
+
+/// This takes a token that refers to a variable and it will return the token
+/// to the expression that the variable is assigned to. If its not valid to
+/// make such substitution then it will return the original token.
+const Token * followVariableExpression(const Token * tok, bool cpp)
+{
+    if(!tok)
+        return tok;
+    // Skip array access
+    if(Token::Match(tok, "%var% ["))
+        return tok;
+    // Skip following variables if it is used in an assignment
+    if(Token::Match(tok->astParent(), "%assign%") || Token::Match(tok->next(), "%assign%"))
+        return tok;
+    const Variable * var = tok->variable();
+    const Token * varTok = getVariableInitExpression(var);
+    if(!varTok)
+        return tok;
+    // Skip array access
+    if(Token::simpleMatch(varTok, "["))
+        return tok;
+    if(var->scope() != tok->scope() && !var->isConst())
+        return tok;
+    if(var->isStatic() && !var->isConst())
+        return tok;
+    if(var->isArgument())
+        return tok;
+    // If this is in a loop then check if variables are modified in the entire scope
+    const Token * endToken = isInLoopCondition(tok) ? tok->scope()->bodyEnd : tok;
+    const Token * startToken = varTok;
+    while(Token::Match(startToken, "%op%|.|(|{"))
+        startToken = startToken->astOperand1();
+    // Skip if the variable its referring to is modified
+    for(const Token * tok2 = startToken;tok2 != endToken;tok2 = tok2->next()) {
+        if(Token::simpleMatch(tok2, ";"))
+            break;
+        if (tok->tokType() == Token::eIncDecOp || 
+            tok->isAssignmentOp() || 
+            Token::Match(tok2, "* %var%") || 
+            Token::Match(tok2, "%name% .|[|++|--|%assign%")) {
+            return tok;
+        }
+        
+        if(Token::Match(tok2, "%var%") && isVariableChanged(tok2, endToken, tok2->varId(), false, nullptr, cpp)) {
+            return tok;
+        }
+    }
+    if (!isVariableChanged(varTok, endToken, tok->varId(), false, nullptr, cpp))
+        return varTok;
+    return tok;
+}
+
 bool isSameExpression(bool cpp, bool macro, const Token *tok1, const Token *tok2, const Library& library, bool pure)
 {
-    if (tok1 == nullptr && tok2 == nullptr)
+    if (tok1 == nullptr && tok2 == nullptr) 
```

Comment:
I suggest that you install astyle, version 3.0.1. then run the runastyle script in the cppcheck root folder. that takes care of trailing spaces and other stuff.

Reply:
I am using ubuntu which comes with astyle 2.05.1. I tried downloading it and building it, but the astyle project is a complete mess.

Reply:
I use debian and also have to download and build the source code. Yes the compilation is "unusual". I guess cppcheck compilation is "unusual" also.

I use something like:

    tar xzvf astyle_3.0.1_linux.tar.gz
    cd astyle/build/gcc
    make && sudo make install

It installs astyle in /usr/bin

Reply:
> I guess cppcheck compilation is "unusual" also.

Cppcheck compilation is not unusual at all. In fact, I can install the latest with just `cget install danmar/cppcheck`.

> It installs astyle in /usr/bin

I would prefer to install it to a local directory. Let me look into this.